rust/kernel/thread.rs

```diff
@@ -0,0 +1,198 @@
+
+use crate::c_types;
+use crate::error::{ptr_to_result, Error, KernelResult};
+use crate::{bindings, CStr};
+
+use alloc::boxed::Box;
+use core::ops::FnOnce;
+
+extern "C" {
+    #[allow(improper_ctypes)]
+    fn rust_helper_get_task_struct(task: *mut bindings::task_struct);
+    #[allow(improper_ctypes)]
+    fn rust_helper_put_task_struct(task: *mut bindings::task_struct);
+}
+
+/// Function passed to `kthread_create_on_node` as function pointer. No other user.
+#[no_mangle]
+unsafe extern "C" fn rust_thread_func(data: *mut c_types::c_void) -> c_types::c_int {
+    // ·Box::from_raw()` to get the ownership of the closure.
+    let c = Box::from_raw(data as *mut Box<dyn FnOnce() -> KernelResult<()>>);
+
+    let ret = c();
+
+    match ret {
+        Ok(_) => 0,
+        Err(e) => e.to_kernel_errno(),
+    }
+}
+
+/// A kernel thread handle
+pub struct Thread {
+    /// Pointer to kernel thread
+    task: *mut bindings::task_struct,
+}
+
+impl Thread {
+    /// Create a new thread
+    ///
+    /// This function might sleep due to the memory allocation and waiting for completion in
+    /// `kthread_create_on_node`. Therefore cannot call this in a atomic context.
+    ///
+    /// # Examples:
+    /// use kernel::thread::Thread;
+    /// use alloc::boxed::Box;
+    ///
+    /// let mut a = 1;
+    ///
+    /// let t = Thread::new(
+    ///   move || {
+    ///     let b = Box::try_new(42)?;
+    ///
+    ///     for _ in 0..10 {
+    ///         a = a + 1;
+    ///         println!("Hello Rust Thread {}", a + b.as_ref());
+    ///     }
+    ///     Ok(())
+    ///   },
+    ///   cstr!("rust-thread")
+    /// )?;
+    ///
+    /// t.wake_up();
+    pub fn new<F>(f: F, name: CStr) -> KernelResult<Self>
+    where
+        F: FnOnce() -> KernelResult<()>,
+        F: Send + 'static,
+    {
+        // Allocate closure here, because this function maybe returns before `rust_thread_func`
+        // (the function that use the closure) get executed.
+        let bf: Box<dyn FnOnce() -> KernelResult<()> + 'static> = Box::try_new(f)?;
+
+        // Double boxing here because `dyn FnOnce` is a fat pointer, and we can only pass a usize
+        // as the `data` for `kthread_create_on_node`.
+        //
+        // We `into_raw` from this side, and will `from_raw` at the other side to transfer the
+        // ownership of the boxed data.
+        let db = Box::into_raw(Box::try_new(bf)?) as *mut _;
+
+        let task;
+
+        // SAFETY:
+        //
+        // * `db` is a proper pointer (generated by `Box::into_raw()`), and if succeed, the new
+        // thread will get the ownership.
+        //
+        // * `kthread_create_on_node` will copy the content of name, so we don't need to make the
+        // name live longer.
+        unsafe {
+            task = bindings::kthread_create_on_node(
+                Some(rust_thread_func),
+                db,
+                bindings::NUMA_NO_NODE,
+                "%s".as_ptr() as _,
```

Comment:
This needs to be a CStr as well as it is not guaranteed to be null-terminated. (It may happen to be now by chance, but it's not guaranteed.) The `cstr!` would help you here.

Reply:
Done in v3